administration/src/cards/CardBlueprint.ts

```diff
@@ -5,144 +5,205 @@ import { CardExtensions, CardInfo } from '../generated/card_pb'
 const MAX_NAME_LENGTH = 30
 // Due to limited space on the qr code
 const MAX_ENCODED_NAME_LENGTH = 50
 
-export type JSONCardBlueprint<E = ExtensionInstance> = {
+export type CardBlueprint = {
   id: number
   fullName: string
-  expirationDate: string | null
-  extensions: (E extends Extension<infer T, unknown> ? JSONExtension<T> : never)[]
+  // TODO null necessary? csv?
+  expirationDate: number | null
+  extensions: Partial<ExtensionState>
 }
 
-/**
- * Blueprint for a new card. This object contains data about a future card, which will be created.
- */
-export class CardBlueprint {
-  id: number
-  fullName: string
-  expirationDate: PlainDate | null
-  extensions: ExtensionInstance[]
-  cardConfig: CardConfig
-
-  constructor(fullName: string, cardConfig: CardConfig, initParams?: Parameters<CardBlueprint['initialize']>) {
-    this.cardConfig = cardConfig
-    this.fullName = fullName
-    this.expirationDate = initParams ? PlainDate.fromLocalDate(new Date()).add(cardConfig.defaultValidity) : null
-    this.extensions = cardConfig.extensions.map(Extension => new Extension())
-    this.id = Math.floor(Math.random() * 1000000) // Assign some random ID
-    if (initParams) {
-      this.initialize(...initParams)
-    }
+type CardBlueprintKey = keyof Omit<Omit<CardBlueprint, 'extensions'>, 'id'>
+type KeyType = CardBlueprintKey | ExtensionKey | null
+
+const createRandomId = () => Math.floor(Math.random() * 1000000)
+
+export const initializeCardBlueprint = (
+  cardConfig: CardConfig,
+  region: Region | undefined = undefined,
+  { id, fullName, expirationDate, extensions }: Partial<CardBlueprint> = {}
+): CardBlueprint => {
+  // TODO Why only if region is defined?
+  // const defaultExpirationDate = region ? PlainDate.fromLocalDate(new Date()).add(cardConfig.defaultValidity).toDaysSinceEpoch() : null
+  const defaultExpirationDate = PlainDate.fromLocalDate(new Date()).add(cardConfig.defaultValidity).toDaysSinceEpoch()
+  const defaultExtensions = cardConfig.extensions.reduce(
+    (acc, extension) => ({
+      ...acc,
+      ...extension.getInitialState(extension.name === REGION_EXTENSION_NAME ? region : undefined),
+    }),
+    {}
+  )
+
+  return {
+    id: id ?? createRandomId(),
+    fullName: fullName ?? '',
+    extensions: {
+      ...defaultExtensions,
+      ...(extensions ?? {}),
+    },
+    expirationDate: expirationDate ?? defaultExpirationDate,
   }
+}
 
-  setValue(key: string, value: string): void {
-    switch (key) {
-      case this.cardConfig.nameColumnName:
-        this.fullName = value
-        break
-      case this.cardConfig.expiryColumnName:
-        this.setExpirationDate(value)
-        break
-      default:
-        const extensionIdx = this.cardConfig.extensionColumnNames.indexOf(key)
-        if (extensionIdx === -1) {
-          return
-        }
-        this.extensions[extensionIdx].fromString(value)
-    }
-  }
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
```

Comment:
I was not able to construct the correct type, feel free to try :/

Reply:
hm i removed the comment and don't get any error

Reply:
i think this is the correct type:
`Array<{
  extension: Extension<InferExtensionStateType<typeof Extensions[number]>>;
  state: InferExtensionStateType<typeof Extensions[number]>;
}>`

Reply:
Thank you @ztefanie, nice! :tada: